src/main/scala/chisel3/util/experimental/BarrelShifter.scala

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.util.experimental
+
+import chisel3._
+import chisel3.util.{Mux1H, UIntToOH}
+
+/** A Barrel Shifter implementation for Vec type.
+  *
+  * @param inputs           input signal to be shifted, should be a [[Vec]] type.
+  * @param shiftInput       input signal to indicate the shift number, encoded in UInt.
+  * @param shiftType        [[ShiftType]] to indicate the type of shifter.
+  * @param shiftGranularity how many bits will be resolved in each layer.
+  *                         For a smaller `shiftGranularity`, latency will be high, but area is smaller.
+  *                         For a large `shiftGranularity`, latency will be low, but area is higher.
+  *
+  * @example {{{
+  * val input = 0xAABBCCDD.U(32.W)
+  * val bytes = VecInit(input.asBools.grouped(8).map(VecInit(_).asUInt).toSeq)
+  * val shamt = 1.U(2.W)
+  * val output = BarrelShifter.leftRotate(bytes, shamt).asUInt // big endian
+  * // output = 0xDDAABBCC.U(32.W)
+  * // The depth of Mux is 2, whereas originally
+  * // the depth of Mux is log(32) = 5
+  * val output = BarrelShifter.leftRotate(bytes, shamt, 2).asUInt // depth is 1 now
+  * }}}
+  */
+object BarrelShifter {
+  trait ShiftType
+
+  object LeftShift extends ShiftType
+
+  object RightShift extends ShiftType
+
+  object LeftRotate extends ShiftType
+
+  object RightRotate extends ShiftType
+
+  def apply[T <: Data](inputs: Vec[T], shiftInput: UInt, shiftType: ShiftType, shiftGranularity: Int = 1): Vec[T] = {
```

Comment:
`shiftGranularity` should GT 0, you should add `require` here

Reply:
Added.